File: apps/api/src/modules/proxy/content-analyzer.ts
import { createId } from "@paralleldrive/cuid2";

export interface ContentAnalysis {
  readonly hasImages: boolean;
  readonly imageCount: number;
  readonly hasToolUse: boolean;
  readonly toolCount: number;
  readonly toolNames: readonly string[];
  readonly sessionId: string | null;
}

const countImages = (messages: unknown[]): number => {
  let count = 0;

  for (const message of messages) {
    const msg = message as Record<string, unknown>;
    const content = msg.content;
    if (!Array.isArray(content)) continue;

    for (const block of content) {
      const b = block as Record<string, unknown>;
      // OpenAI: type: "image_url"
      if (b.type === "image_url") count++;
      // Anthropic: type: "image"
      if (b.type === "image") count++;
    }
  }

  return count;
};

const extractToolNames = (tools: unknown[]): readonly string[] =>
  tools.flatMap((tool) => {
    const t = tool as Record<string, unknown>;
    // OpenAI: tools[].function.name
    const fn = t.function as Record<string, unknown> | undefined;
    if (fn && typeof fn.name === "string") return [fn.name];
    // Anthropic: tools[].name
    if (typeof t.name === "string") return [t.name];
    return [];
  });

const extractFunctionNames = (functions: unknown[]): readonly string[] =>
  functions.flatMap((fn) => {
    const f = fn as Record<string, unknown>;
    return typeof f.name === "string" ? [f.name] : [];
  });

export const analyzeContent = (
  body: unknown,
  sessionHeader?: string | null
): ContentAnalysis => {
  const empty: ContentAnalysis = {
    hasImages: false,
    hasToolUse: false,
    imageCount: 0,
    sessionId: null,
    toolCount: 0,
    toolNames: []
  };

  if (!body || typeof body !== "object") return empty;

  const b = body as Record<string, unknown>;

  // Count images in messages
  const imageCount = Array.isArray(b.messages) ? countImages(b.messages) : 0;

  // Detect tool use: OpenAI uses "tools" or "functions", Anthropic uses "tools"
  const toolInfo = Array.isArray(b.tools)
    ? {
        hasToolUse: true,
        toolCount: b.tools.length,
        toolNames: extractToolNames(b.tools)
      }
    : Array.isArray(b.functions)
      ? {
          hasToolUse: true,
          toolCount: b.functions.length,
          toolNames: extractFunctionNames(b.functions)
        }
      : { hasToolUse: false, toolCount: 0, toolNames: [] as readonly string[] };

  // Extract session ID from header or body metadata
  const sessionId = (() => {
    if (sessionHeader) return sessionHeader;
    const metadata = b.metadata as Record<string, unknown> | undefined;
    if (metadata && typeof metadata.session_id === "string") {
      return metadata.session_id;
    }
    return createId();
  })();

  return {
    hasImages: imageCount > 0,
    ...toolInfo,
    imageCount,
    sessionId
  };
};

File: apps/api/src/proxy/content_analyzer.py
from typing import Any


def analyze_content(body: dict[str, Any], session_header: str | None = None) -> dict:
    messages = body.get("messages", [])
    has_images = False
    image_count = 0
    has_tool_use = False
    tool_count = 0
    tool_names: list[str] = []

    for msg in messages:
        content = msg.get("content", "")
        if isinstance(content, list):
            for part in content:
                if isinstance(part, dict) and part.get("type") == "image_url":
                    has_images = True
                    image_count += 1

        if msg.get("role") == "assistant" and msg.get("tool_calls"):
            has_tool_use = True
            for tc in msg["tool_calls"]:
                tool_count += 1
                fn = tc.get("function", {})
                name = fn.get("name")
                if name:
                    tool_names.append(name)

    tools = body.get("tools", [])
    if tools:
        has_tool_use = True
        tool_count = max(tool_count, len(tools))

    session_id = session_header or body.get("metadata", {}).get("session_id")

    return {
        "has_images": has_images,
        "image_count": image_count,
        "has_tool_use": has_tool_use,
        "tool_count": tool_count,
        "tool_names": tool_names,
        "session_id": session_id,
    }
